### PR TITLE
Enhance SQ_WEIGHT

### PR DESCRIPTION
**Pull request by PhoenixWorthVCD**
## Description

Enhance the existing SQ_WEIGHT algorithm that consists of skipping the evaluation of NSQ blocks (HA, HB, VA, VB, H4 and V4 shapes) based on the relative cost of SQ, H, and V shapes.  Skipping is determined based on the following criterion:

 - skip HA, HB and H4 if (valid SQ and H) and (H_COST > (SQ_WEIGHT * SQ_COST) / 100) 
 - skip VA, VB and V4 if (valid SQ and V) and (V_COST > (SQ_WEIGHT * SQ_COST) / 100)

Now, the scaling factor SQ_WEIGHT (that determines how likely it is to skip blocks) is a function of the qp, block shape, prediction mode, block coeffs, and encode mode.  In summation, this PR adds the ability to refine SQ_WEIGHT based on the aforementioned block properties, whereas previously the SQ_WEIGHT was a constant value determined only by the encode mode.

The SQ_WEIGHT is determined based on the following tables.

Preset(s) | BASE 
-- | --
MR | ∞
M0 | 105
M1 | 105
M2 | 100
M3 & beyond | 95

The below listed SQ_WEIGHT modifiers are the same for H and V blocks.

Block Property | OFFSET
--|--
QP 20| -10
H4 | +5
HA/B & INTRA | +5
HA/B & no coeffs|-10

SQ_WEIGHT = BASE + OFFSET

## Author

@PhoenixWorthVCD 
@hguermaz 

## Type of Change

Enhancement

## Testing and Performance

The data in the first table is based on the objective-1-fast dataset (8 bit).

| |BDR | Speed Dev.|
--|--|--
|M0|0.07% | 11.2%|
|M1|0.08% | 10.3%|
|M2|0.21% | 20.4%|

The below 10 bit data is based on  fourteen 720p true 10 bit clips.

| |BDR | Speed Dev.|
--|--|--
|M0|0.07% | 13.0%|
|M1|0.02% | 10.4%|
|M2|0.20% | 19.5%|

### CODE REVIEW

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c [hunk 1]
```diff
@@ -6272,10 +6272,120 @@ void md_encode_block(SequenceControlSet *scs_ptr, PictureControlSet *pcs_ptr,
...
+uint8_t update_skip_nsq_shapes(SequenceControlSet *scs_ptr, PictureControlSet *pcs_ptr,
+                               ModeDecisionContext *context_ptr) {
+    uint8_t skip_nsq = 0;
+    uint32_t sq_weight = context_ptr->sq_weight;
+
+    // return immediately if the skip nsq threshold is infinite
+    if (sq_weight == (uint32_t)~0) return skip_nsq;
+
+    // use an aggressive threshold for QP 20
+    if (scs_ptr->static_config.qp <= 20) sq_weight -= 10;
```
Comment (mafonsoN) [inline Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: Do we want to have these constants defined somewhere instead? -10, +5, QP 20.

Comment (hguermaz) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: It makes sense. @PhoenixWorthVCD could you please CONSERVATIVE_OFFSET_0, CONSERVATIVE_OFFSET_1, .. and AGRESSIVE_OFFSET_0, ..

Comment (hguermaz) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: It makes sense. @PhoenixWorthVCD could you please CONSERVATIVE_OFFSET_0, CONSERVATIVE_OFFSET_1, .. and AGRESSIVE_OFFSET_0, ..

Comment (hguermaz) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: It makes sense. @PhoenixWorthVCD could you please CONSERVATIVE_OFFSET_0, CONSERVATIVE_OFFSET_1, .. and AGRESSIVE_OFFSET_0, ..

Comment (hguermaz) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: It makes sense. @PhoenixWorthVCD could you please CONSERVATIVE_OFFSET_0, CONSERVATIVE_OFFSET_1, .. and AGRESSIVE_OFFSET_0, ..

Comment (hguermaz) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: It makes sense. @PhoenixWorthVCD could you please CONSERVATIVE_OFFSET_0, CONSERVATIVE_OFFSET_1, .. and AGRESSIVE_OFFSET_0, ..

Comment (hguermaz) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: It makes sense. @PhoenixWorthVCD could you please CONSERVATIVE_OFFSET_0, CONSERVATIVE_OFFSET_1, .. and AGRESSIVE_OFFSET_0, ..

**Comment by hassount** [comment]
@PhoenixWorthVCD can you please rebase this PR and fix the conflicts?

**Comment by mafonsoN** [comment]
Can we also add/update the documentation?

**Comment by hguermaz** [comment]
@mafonsoN https://gitlab.com/intel-svt/svt-av1/-/wikis/Appendix%20SQ%20Weight

**Pull request merged**